### PR TITLE
Added display name format extension point #2507

### LINKED ISSUE #2507: Allow a custom function to name a test
Right now we allow users to append tags to tests names, which is fine - but it would be even better to provide users an extension point to modify the name of the tests based on the test itself and tags (and perhaps other information).

Given a test with tags `Slow` and `NickNack`

status quo:
```
TestClass.nicknack does some stuff[tags = Slow, NickNack]
```

new hotness, in kotlin pseudocode:
```kotlin
val Slow = Tag("Slow")

val testNamer = object : TestNameExtension {
  override fun nameTest(testCase: TestCase, tags: Set<Tag>): String {
    if (tags.contains(Slow)) {
      return "Slow: ${testCase.name} [tags=${(tags - Slow).joinToString()}]"
   }
   return testCase.name
```

would give us a nice test name like 

```
Slow: TestClass.nicknack does some stuff [tags = NickNack]
```

### PR DESCRIPTION
**Pull request by sksamuel**
Closes #2507

### CODE REVIEW

**Comment by jschneidereit** [comment]
Does the append tags implementation require an update after this?

File: kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/extensions/DisplayNameFormatterExtension.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+   /**
+    * Returns a formatted name for a test.
+    */
+   fun format(testCase: TestCase): String
+
+   /**
+    * Returns a formatted name for a spec class.
+    */
+   fun format(kclass: KClass<*>): String
```
Comment (jschneidereit) [inline kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/extensions/DisplayNameFormatterExtension.kt:23]: Does this work for js and native?

**Review by jschneidereit — APPROVED** [review]

File: kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/extensions/DisplayNameFormatterExtension.kt (hunk 1 above)
Comment (sksamuel) [reply kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/extensions/DisplayNameFormatterExtension.kt:23]: Yes

**Comment by sksamuel** [comment]
> Does the append tags implementation require an update after this?

The default formatter appends tags in the old way.

**Pull request merged**